### ISSUE TITLE
Character encoding issues

### ISSUE THREAD

**Issue by jsmall-zzz** [shader-slang/slang#1954]
There appear to be a few issues around what could probably be summarized as 'character encoding'.

It's worth noting that Slang does *not* use character encoding/decoding functionality to read source - the source is read as a stream of bytes, that is assumed to be utf8. Unfortunately there is no checking that this is the case (by looking for a marker), or moving past the marker if there is one. We could work around this by removing the marker, decoding to utf8 and putting in a new blob. The decoding is fairly simple because we don't need to worry about line endings, and we have the 'whole file' do no buffering problems. Then the SourceLoc, file offsets, become the offset into the *marker stripped utf8 cr only* encoded source - not the original source file. Removing ambiguity around line ending encoding means that a module can be used when source encoding may have changed - for example having the same source referenced by the SourceLoc but differently encoded, as can happen with source control systems such as git.

This raises a question about the responsibility of having source in this format. At the moment the FileSystem interface can read a blob, and that is assumed to be an exact binary representation. An implementation could know that source is in the 'canonical source format' or have a faster/better implementation to convert into that canonical format. We don't want to require an implementation to have to implement this though, as it adds significant complexity to implement from scratch. So if a request is made that it can't do it could return 'not implemented' and be read as binary and converted. How to add this functionality? Having an interface we can query for that allows for passing in an enum for read/write perhaps. Perhaps this becomes the main interface - doing so would break binary compatibility. It's also worth noting that having a requirement for 'canonical' source encoding could introduce some other requirements such as 0 termination to improve/simplify lexing.

The StreamReader does handle the encoding of characters that can occur across a buffer boundary. Doing so requires decoding to have to do a fair amount of work checking and refilling the buffer as bytes are read. When the end of the buffer is reached if the encoding is incorrect the reads will append 0 bytes. This is arguably not correct (outputting an invalid character might be better for example). As it stands it also creates a problem because there isn't an easy way for the buffer refill to report an error, so for now if there is an error on stream read it's just ignored.

The StreamWriter assumes the slices written to it always contain 'complete' codepoints. If that's not he case the the conversion will not work correctly, as conversion only works with complete code points. It's perhaps not completely unreasonable to require complete codepoints in a slice.

That both Reader/Writer work working through code point conversion. For the common case of reading/writing utf8/ascii this is a lot of extra work. The Reader/Writer are also complicated by their handling of line endings. 

It seems like a lot of extra complexity/work to make reading/writing test files/intermediates do this work. One way to side step some of these issues would be to have the base Reader/Writer support line ending issues of utf8. 

That any conversions are handled by reading the whole stream, and converting if necessary. This makes the conversion much faster/simpler. 

It may be worth specifying the line ending of the stream (if known) as well as the line ending desired (including leaving as is). This could be handled via something like

```
static SlangResult StreamReader::create(Stream* stream, const ReadOptions& options, RefPtr<TextReader>& outReader);
static SlangResult StreamWriter::create(Stream* stream, const WriteOptions& options, RefPtr<TextWrite>& outWriter);
```

Doing do so allows for different implementations being constructed depending on what is needed, and can do encoding conversion if necessary before construction. 

An alternative would be to have a 'Stream-like' small buffered version, and make it possible to compose the two. For example a non utf8 reader, would be the combination of 'non utf8' stream reader, wrapped in a Cr handler (if necessary). There would also be a need for utf8 decoding that could determine if it doesn't have enough bytes. Having this would avoid the decoder having to be templated on a reader (which currently takes an unused index). Whilst this would have certain advantages, the simpler approach is probably better for our purposes, and the simplest is to just support utf8 with cr/lf handling.

As a small related point when writing intermediates Slang does *not* use the Writer or even stream mechanism, which it probably should.

Conclusion:

* We should fix the handling of utf8 in source - by removing marker and ensuring encoding is utf8
* Ideally we should streamline TextReader/Writer for simplicity and performance

**Comment by natduca** [comment]
Putting with other robustness topics in Q3

**Comment by swoods-nv** [comment]
Fixed by #4135
